### PR TITLE
Feature/sleep class

### PR DESCRIPTION
**Pull request by dietza**
### What does this PR do?
Builds out structure for the Sleep class in Sleep.js including intended methods (see issues).
Adds and passes all tests in Sleep-test.js.

### Where should the reviewer start?
Line 1 of Sleep.js
Line 1 of Sleep-test.js

### How is this manually tested?
run `npm test` in terminal to confirm that tests pass as expected.

### Any relevant tickets?
#6 #16

### CODE REVIEW

File: src/Sleep.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    const nightlyData = this.sleepData.filter(sleepDataOb => {
+      return sleepDataOb.date === date
+    });
+    const sortedByHours = nightlyData.sort((a, b) => {
+      return b.hoursSlept - a.hoursSlept
+    });
+    const winnerID =  sortedByHours[0].userID;
+    const winnerHours = sortedByHours[0].hoursSlept;
+    return { [winnerID] : winnerHours }
+  }
```
Comment (DrewBradley) [inline src/Sleep.js:99]: This is such an amazing and concise method. Great job!

File: src/Sleep.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    const userList = this.sleepData.map(info => info.userID);
+    const totalUsers = [...new Set(userList)];
+    const totalSleepData = this.sleepData.reduce((acc, initial) => {
+      acc += initial[dataMetric]
+      return acc;
+    }, 0)
+    return Math.round((totalSleepData / totalDates.length) / totalUsers.length);
+  }
+
+  findBestQualitySleepers(date) {
```
Comment (DrewBradley) [inline src/Sleep.js:62]: We should consider refactoring this, then refactoring that.

**Review by DrewBradley — APPROVED** [review]
Thank you for adding all those semicolons!

File: src/Sleep.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    const currentUserData = this.filterDataByUser(userID);
+    const startDate = currentUserData.find(info => info.date === date);
+    const startIndex = currentUserData.indexOf(startDate);
+    const endIndex = startIndex + 7;
+    // RETURN TO FIX ENDINDEX CALUCULATION
+    const weeklyDates = currentUserData.slice(startIndex, endIndex);
+    // CHECK WEEKLYDATES IS RETURNING EXPECTED
+    const weeklySleepData = weeklyDates.map(info => info[dataMetric])
+    return weeklySleepData;
+  }
```
Comment (dietza) [inline src/Sleep.js:48]: Saw this on after the branch had been merged - let's remember to come back and fix this method so we can lose the comments (lines 43, 45)!

File: src/Sleep.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    const currentUserSleepData = [];
+    const total = this.sleepData.reduce((sum, sleepInfo) => {
+      if (sleepInfo.userID === userID) {
+        currentUserSleepData.push(sleepInfo)
+        sum += sleepInfo[dataMetric];
+      }
+      return sum;
+    }, 0);
+    return Math.round(total / currentUserSleepData.length);
+  }
```
Comment (dietza) [inline src/Sleep.js:16]: Can we make this more dynamic by taking in the data Array we'll want to refer to?
It looks like we had to do something similar on lines 76 - 80

**Review by dietza — COMMENTED** [review]
Added a couple more comments on a brief review but this was after the PR had already been merged - we may want to come back to fix a couple of these things sooner than later.
#26

**Pull request merged**